plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -590,24 +597,89 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        ObjectMetadata metadata = getS3ObjectMetadata(path);
+        if (metadata == null) {
+            // the bucket root does not exist
+            if (path.getName().isEmpty()) {
+                return false;
+            }
+
+            // check if this path is a directory
+            Iterator<LocatedFileStatus> iterator = listPath(path, OptionalInt.of(1), ListingMode.SHALLOW_ALL);
+            if (!iterator.hasNext()) {
+                return false;
             }
         }
-        catch (FileNotFoundException e) {
-            return false;
+        else {
+            // the bucket root requires special handling
+            if (!path.getName().isEmpty()) {
+                MediaType metadataContentType = MediaType.parse(metadata.getContentType());
+                if (!metadataContentType.is(DIRECTORY_MEDIA_TYPE)) {
+                    if (!listPath(path, OptionalInt.of(1), ListingMode.SHALLOW_ALL).hasNext()) {
+                        // delete the object
+                        return deleteObject(keyFromPath(path));
+                    }
+
+                    // Delete the object which has a key corresponding to the prefix to be deleted
+                    deleteObject(keyFromPath(path));
+                }
+            }
         }
 
         if (!recursive) {
             throw new IOException("Directory " + path + " is not empty");
         }
 
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        String bucketName = getBucketName(uri);
+        Iterator<S3ObjectSummary> listings = S3Objects.withPrefix(s3, bucketName, keyFromPath(path) + PATH_SEPARATOR).iterator();
```

Comment:
My fault here for using `S3Objects.withPrefix` in an attempt to be concise in my suggested code, we probably want to add a new `ListMode` enum for `KEYS_WITH_PREFIX` and switch this to use `listPath(path, OptionalInt.empty(), KEYS_WITH_PREFIX)` so that we reuse the existing implementation and get metrics like the number of S3 listing calls. I recommend the name `KEYS_WITH_PREFIX` because `RECURSIVE_ALL` would suggest that all "directories" be present in the enumeration, but they won't be- all _keys_ will be returned, but no "directories" that are themselves just a prefix to some child key.

For example, say we have keys:
- `s3://bucket/path/key-1`
- `s3://bucket/path/subpath/key-2`

`listPath("s3://bucket/", OptionalInt.empty(), RECURSIVE_ALL)` would be expected to return:
- `s3://bucket/path`
- `s3://bucket/key-1`
- `s3://bucket/subpath`
- `s3://bucket/subpath/key-2`

Whereas `listPath("s3://bucket/", OptionalInt.empty(), KEYS_WITH_PREFIX)` will only return the actual keys above.

Reply:
Thank you for pointing out `listPath` code reuse.

`RECURSIVE_ALL` does not exist. 

However we already have `RECURSIVE_FILES_ONLY` which very likely fits this specific use case.

Reply:
Correct, `RECURSIVE_ALL` does not exist but would be the missing "quadrant" in the space of `(SHALLOW | RECURSIVE) * (ALL | FILES_ONLY)`. The implied semantics of such a mode would be to include intermediate "directories" that have no corresponding object/key, which isn't really what we care about for deletes or renames. 

So the proposed new mode should be `RECURSIVE` but _not_ filter out "directories" which have actual S3 object keys (either marked with the `application/x-directory` content-type or ending with the suffix `"_$folder$"`), which `RECURSIVE_FILES_ONLY` does [here](https://github.com/trinodb/trino/blob/f6cad5e9450df12ba9e10071522587b09abde7b7/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java#L694,L697). Instead, I propose the name of that new listing mode emphasize that it lists _all keys with a given prefix_ which will include "directories", but not "directories" that have no corresponding S3 object.

Reply:
@pettyjamesm  I follow the logic of the above statement.

However, on testing I noticed that that `RECURSIVE_FILES_ONLY` does  fit the requirements of this method.

See the tests:
- `testDeleteNonRecursivelyPrefixContainingMultipleObjectsShouldFail`
- `testDeleteRecursivelyPathContainingPrefixCorrespondingAlsoToAnObjectKey`

Reply:
After discussing on Slack with @pettyjamesm I understand now that we can have:
- objects
- directory objects (with content type `application/x-directory`)
- virtual directories (which exist only because they hold an object or a directory object)

I followed the hints from @pettyjamesm  and refactored the PR to be able to deal with directory objects.